packages/application-shell/src/configure-apollo.ts

```diff
@@ -100,22 +100,9 @@ export const createApolloClient = () =>
       fragmentMatcher,
       dataIdFromObject: result => {
         const customResult = result as CustomIdGetterObj;
-        if (
-          !customResult.id &&
-          customResult.key &&
-          customResult.__typename &&
-          typeNamesWithoutIdAsIdentifier.includes(customResult.__typename)
-        )
-          return `${customResult.__typename}:${customResult.key}`;
```

Comment:
the new condition will return the same value, unless :
`!customResult.id ` ✔︎
`customResult.key` ✔︎
`customResult.__typename ` ✘

In this case the old condition will not fire and we'll move to the else and  return
either `${customResult.__typename}:undefined`; for Reference type or
just `undefined` if it wasn't  a Reference

Reply:
Ah, you're saying with the simplifcation we actually cover edge cases better? Sorry Monday morning slowness :)

Reply:
> you're saying with the simplifcation we actually cover edge cases better? 


I think so. I was skeptical at first, but the only possible downside that I could find was if the `__typename` is not available.

I'm not sure if this is possible, but in this case, the proposed implementation will not do better or worse.
Maybe if this is possible, we can also default to e.g `---` so the cache id, in this case, would be

`---: resource-id`, instead of `undefined: resource-id`

Reply:
I looked at the previous commits and PRs and I think this change should be fine. My question would be just the precedence. So always id if possible otherwise key or key if present.

Reply:
The most important thing for the cache id is that it must be unique. Given that, I think the `id` should have higher precedence.